### DISCUSSION TITLE
./nnn: symbol lookup error: ./nnn: undefined symbol: COLORS

### DISCUSSION THREAD

**Discussion by mikarm999** [jarun/nnn#1687] [Q&A] [answered]
Hi,
When compiling nnn-4.8 from source, I did make and it passed without errors.
However when I run nnn, the following error appears:
./nnn: symbol lookup error: ./nnn: undefined symbol: COLORS
Could you please help me?
My operation system is OpenSuse Tumbleweed.
Thank you!

**Comment by jarun** [comment]
`COLORS` and `COLOR_PAIRS` should come from ncurses: https://linux.die.net/man/3/color_pair
Please refer tot he section _Routine Descriptions_.

So it seems your terminal does not have color capabilities.

Try:

    export NO_COLOR=1
    unset NNN_COLORS

**Reply by mikarm999** [reply]
Thank you for your message. I use alacrity, which has color capabilities, and the same error appears with other terminals.
At the same time if I use static version nnn-static-nerd , this problem does not appear.

**Comment by N-R-K** [comment]
>  When compiling nnn-4.8 from source, I did make and it passed without errors.

Show the full output of `make` as well as the command you used to invoke `make`. Also include if you've modified the Makefile or not.

**Reply by mikarm999** [reply]
Thank you for the prompt reply. I did the following:
git clone https://github.com/jarun/nnn.git
cd nnn
make
The output was: 
cc  -std=c11 -Wall -Wextra -Wshadow -O3 -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -I/home/prince/.linuxbrew/Cellar/ncurses/6.4/include/ncursesw -I/home/prince/.linuxbrew/Cellar/ncurses/6.4/include  -o nnn  src/nnn.c -lreadline -L/home/prince/.linuxbrew/Cellar/ncurses/6.4/lib -lncursesw -lpthread
Then,
./nnn
./nnn
./nnn: symbol lookup error: ./nnn: undefined symbol: COLORS

Thank you very much!

**Reply by jarun** [reply]
Does it work if you do a [static build](https://github.com/jarun/nnn/wiki/Developer-guides#static-compilation)?

    make O_NERD=1 O_STATIC=1 strip

I feel it may be linking to the wrong ncurses library during execution.

**Reply by mikarm999** [reply]
The output of 
make O_NERD=1 O_STATIC=1 strip
is
cc -DNORL -DNERD -DICONS_INCLUDE=\"icons-generated-nerd.h\" -std=c11 -Wall -Wextra -Wshadow -O3 -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -I/home/prince/.linuxbrew/Cellar/ncurses/6.4/include/ncursesw -I/home/prince/.linuxbrew/Cellar/ncurses/6.4/include -static -DICONS_GENERATE -o src/icons-hash-gen src/icons-hash.c
/usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: cannot find -lc: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [Makefile:221: src/icons-generated-nerd.h] Error 1

**Reply by mikarm999** [reply]
Yes, possibly the problem is in homebrew installed and there is a confusion in the libraries.
If I compile as another user which does not have homebrew installed everething is ok:
*************
make
cc  -std=c11 -Wall -Wextra -Wshadow -O3 -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -I/usr/include/ncursesw   -o nnn  src/nnn.c -lreadline -lncursesw -Wl,--push-state,--as-needed -ltinfo -Wl,--pop-state  -lpthread
*************************
make O_NERD=1
cc -DNERD -DICONS_INCLUDE=\"icons-generated-nerd.h\" -std=c11 -Wall -Wextra -Wshadow -O3 -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -I/usr/include/ncursesw   -DICONS_GENERATE -o src/icons-hash-gen src/icons-hash.c
./src/icons-hash-gen > src/icons-generated-nerd.h
cc -DNERD -DICONS_INCLUDE=\"icons-generated-nerd.h\" -std=c11 -Wall -Wextra -Wshadow -O3 -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -I/usr/include/ncursesw   -o nnn  src/nnn.c -lreadline -lncursesw -Wl,--push-state,--as-needed -ltinfo -Wl,--pop-state  -lpthread
*******************
So the question is: how to avoid this confusion?
Thank you.

**Reply by N-R-K** [reply]
> So the question is: how to avoid this confusion?

You'll need to ask that in some macos/homebrew forum.

**Reply by mikarm999** [reply]
Unfortunately, I could not find the answer published somewhere, however I find a way: exclude from the path everything related to homebrew (redefine PATH), then compile, and launch the program, now it works well. 
And it still works when we restore the PATH. So this may be a solution. Maybe we can close the discussion. Thank you!

**Reply by mikarm999** [reply, marked as answer]
Simple script which does the task:
***********************
#!/bin/bash

#O_DEBUG := 0  # debug binary
#O_NORL := 0  # no readline support
#O_PCRE := 0  # link with PCRE library
#O_NOLOC := 0  # no locale support
#O_NOMOUSE := 0  # no mouse support
#O_NOBATCH := 0  # no built-in batch renamer
#O_NOFIFO := 0  # no FIFO previewer support
#O_CTX8 := 0  # enable 8 contexts
#O_ICONS := 0  # support icons-in-terminal
# O_NERD := 0  # support icons-nerdfont
#O_QSORT := 0  # use Alexey Tourbin's QSORT implementation
#O_BENCH := 0  # benchmark mode (stops at first user input)
#O_NOSSN := 0  # enable session support
#O_NOUG := 0  # disable user, group name in status bar

git clone https://github.com/jarun/nnn.git
PATH_OLD = $(echo $PATH)
export PATH="/usr/local/bin/:/usr/bin/:/bin"
cd nnn

make O_CTX8=0 O_NERD=1 O_ICONS= O_NOUG=0 O_QSORT=0 

sudo make uninstall && sudo make install
export PATH=$(echo $PATH_OLD)
exit
